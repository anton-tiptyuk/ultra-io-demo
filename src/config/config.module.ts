import { Module, Global } from '@nestjs/common';
import { TypeOrmModule } from '@nestjs/typeorm';

import { ConfigService } from './config.service';

@Global()
@Module({
  imports: [
    TypeOrmModule.forRootAsync({
      inject: [ConfigService],
      useFactory: async (configService: ConfigService) =>
        configService.getOrmConfig()
    }),
  ],
  providers: [ConfigService],
  exports: [ConfigService],
})
export class ConfigModule { }
