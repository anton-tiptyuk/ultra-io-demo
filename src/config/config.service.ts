import { Injectable } from '@nestjs/common';
import { TypeOrmModuleOptions } from '@nestjs/typeorm';
import * as Joi from '@hapi/joi';
import * as dotenv from 'dotenv';
import * as path from 'path';

import { SnakeNamingStrategy } from '../db/SnakeNamingStrategy';

interface IEnvConfig {
  [key: string]: any;
}

const DATABASE_HOST = 'DATABASE_HOST';
const DATABASE_PORT = 'DATABASE_PORT';
const DATABASE_NAME = 'DATABASE_NAME';
const DATABASE_USERNAME = 'DATABASE_USERNAME';
const DATABASE_PASSWORD = 'DATABASE_PASSWORD';
const DATABASE_SYNCHRONIZE = 'DATABASE_SYNCHRONIZE';
const DATABASE_LOGGING = 'DATABASE_LOGGING';

const DEFAULT_PORT = 3000;
const ext = path.extname(__filename);
const dbDir = path.relative(process.cwd(), path.resolve(`${__dirname}/../db`));

@Injectable()
export class ConfigService {
  private readonly envConfig: IEnvConfig;

  constructor() {
    dotenv.config();
    this.envConfig = ConfigService.validateConfig(process.env);
  }

  private static validateConfig(config: IEnvConfig): IEnvConfig {
    const schema: Joi.ObjectSchema = Joi.object().keys({
      PORT: Joi.number().default(DEFAULT_PORT),

      [DATABASE_PORT]: Joi.number().default(5432),
      [DATABASE_HOST]: Joi.string().required(),
      [DATABASE_NAME]: Joi.string().required(),
      [DATABASE_USERNAME]: Joi.string().required(),
      [DATABASE_PASSWORD]: Joi.string().required(),
      [DATABASE_SYNCHRONIZE]: Joi.boolean().default(false),
      [DATABASE_LOGGING]: Joi.boolean().default(false),
    });

    const { error, value } = schema.validate(config, { allowUnknown: true });

    if (error)
      throw new Error(`Config validation error: ${error.message}`);

    return value;
  }

  get<T>(key: string): T {
    return this.envConfig[key];
  }

  getOrmConfig() {
    return {
      dropSchema: false,
      entities: [`${dbDir}/models/**/*.entity${ext}`],
      migrations: [`${dbDir}/migrations/**/*${ext}`],
      cli: {
        migrationsDir: `${dbDir}/migrations`,
        // entitiesDir: `${dbDir}/models`,
      },
      namingStrategy: new SnakeNamingStrategy(),
      seeds: [`${dbDir}/seeds/**/*.seed${ext}`],
      factories: [`${dbDir}/factories/**/*.factory${ext}`],

      type: 'postgres',
      host: this.get(DATABASE_HOST),
      port: this.get<number>(DATABASE_PORT),
      database: this.get(DATABASE_NAME),
      username: this.get(DATABASE_USERNAME),
      password: this.get(DATABASE_PASSWORD),
      synchronize: this.get<boolean>(DATABASE_SYNCHRONIZE),
      logging: this.get<boolean>(DATABASE_LOGGING),
    } as TypeOrmModuleOptions;
  }

}
