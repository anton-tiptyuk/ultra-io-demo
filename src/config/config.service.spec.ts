import { Test, TestingModule } from '@nestjs/testing';
import { ConfigService } from './config.service';

describe('ConfigService', () => {
  let service: ConfigService;

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      providers: [ConfigService],
    }).compile();

    service = module.get<ConfigService>(ConfigService);
  });

  it('should be defined', () => {
    expect(service).toBeDefined();
  });

  it('should implement .getOrmConfig()', () => {
    expect(typeof service.getOrmConfig()).toEqual('object');
  });

  it('should fail to validate wrong config', () => {
    const v = ConfigService['validateConfig'];
    const validationCall = () => v({ any: 'value' });
    expect(validationCall).toThrowErrorMatchingSnapshot();
  });

});
