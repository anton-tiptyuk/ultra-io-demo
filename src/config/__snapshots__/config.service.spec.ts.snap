// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfigService should fail to validate wrong config 1`] = `"Config validation error: \\"DATABASE_HOST\\" is required"`;
